[Ng Jing Xue] iP

njxue opened the pull request:
# Duke
> _"Woof"_ - Doge

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need is,
1. Download it [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ)
2. Double-click it
3. Add your tasks
4. ***VOILA*** 

AND IT'S 😱😲 **FREEEEEE**  😱😲 

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon) 

A code snippet of `main` from Java:
``` java
public static void main(String[] args) {
    System.out.println("woof");
}
```

Commit:
Add Gradle support

njxue committed:
Add greet user functionality

njxue committed:
Add say bye functionality

njxue committed:
Add query for input functionality

njxue committed:
Add ability to store text entered by user and display them back to user when requested

njxue committed:
Add Task class

njxue committed:
Add getter for Task description

njxue committed:
Add unmark Task method

njxue committed:
Add adding task and listing tasks functionality

njxue committed:
Add ability to mark and unmark tasks as done

njxue committed:
Move all input handling to start() method

njxue committed:
Add toString method to Task

njxue committed:
Add Deadline class

njxue committed:
Add Todo class

njxue committed:
Add Event class

njxue committed:
Add support for trcking different types of tasks

njxue committed:
Add semi-automated tests for Duke

njxue committed:
Implement exception handling for unknown Duke commands

njxue committed:
Implement exception handling for missing description or date in task

njxue committed:
Implement exception handling for incorrect task index

njxue committed:
Add support for deleting tasks from list

njxue committed:
Add more tests with delete command and exception handling

njxue committed:
Add Command enums and refactor exception handling

njxue committed:
Add tests to test all commands and exceptions

njxue committed:
Add ability to save tasks to local storage

njxue committed:
Specify format of each task in the task list in storage

njxue committed:
Add ability to load tasks from hard disk on start up

njxue committed:
Handle case where data file does not exist at start

njxue committed:
Store dates as LocalDate and time as LocalTime instead of strings

njxue committed:
Add default time of 23:59 if time argument is not provided

njxue committed:
Replace LocalDate and LocalTime with LocalDateTime

njxue committed:
Merge branch 'branch-Level-7'

njxue committed:
Fix merge conflict caused by conflicting name for LocalDateTime members

njxue committed:
Fix bug where LocalDateTime of tasks in storage cannot be parsed

njxue committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

njxue committed:
Configure gradle

njxue committed:
Add Ui class which deals with interactions with the user

njxue committed:
Add Storage class which deals with loading tasks from the file and saving tasks in the file

njxue committed:
Add TaskList class which contains the task list and performs operations on the list

njxue committed:
Add Command classes and Parser class

njxue committed:
Add command-specific print methods to Ui class

njxue committed:
Strip white spaces from command and command arguments

njxue committed:
Fix incorrect task index bug

njxue committed:
Make Duke constructor accept path string

njxue committed:
Update exception message when there is an error in writing to tasks storage

njxue committed:
Add success message when tasks load successfully from disk

njxue committed:
Check is nextline exists when using Scanner to support text-ui-test

njxue committed:
Add more tests

njxue committed:
Organize into duke packages

njxue committed:
Update path to Duke file

njxue committed:
Fix Duke not exiting bug

njxue committed:
Add JUnit tests

njxue committed:
Package test files

njxue committed:
Change mainClassName

njxue committed:
Add JavaDocs to all src files

njxue committed:
Edit code to adhere to coding standards

njxue committed:
Add Find command

njxue committed:
Resolve merge conflicts when merging master with A-CodingStandard

njxue committed:
Resolve merge conflict from merging master with branch-Level-9

njxue committed:
Fix wrong class path to Duke

njxue committed:
Add new tests to test the find command

njxue committed:
Remove final keyword

njxue committed:
Add author and version to JavaDoc of class header

njxue committed:
Set members from protected to private

njxue committed:
Capitalise constant fields

njxue committed:
Remove <code></code> tags in JavaDocs

njxue committed:
Add dependencies for Java FX

njxue committed:
Fix style violations with checkstyle

njxue committed:
Fix style violations with checkstyle

njxue committed:
Implement JavaFx without FXML

njxue committed:
Implement JavaFx without FXML

njxue committed:
Configure build for JavaFx

njxue committed:
Change visibilities from public to private

njxue committed:
Merge branch 'branch-A-CheckStyle'

njxue committed:
Implement JavaFX with FXML

njxue committed:
Fix IndexOutOfBoundsExceptio

njxue committed:
Rearrange import statements to adhere to checkstyle

njxue committed:
Rearrange import statements to adhere to checkstyle

njxue committed:
Add and modify JavaDocs

njxue committed:
Configure build for JavaFx

njxue committed:
Improve Ui

njxue committed:
Resolve merge conflict

njxue committed:
Change mainClassName to Launcher

njxue committed:
Merge branch 'branch-Level-10'

dexter-sim commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,31 @@
+     * @param storage Storage object which loads and saves tasks.
+     * @throws DukeException If an exception occurs.
+     */
+    public abstract String execute(TaskList tasks, Ui ui, Storage storage) throws DukeException;
```
I see that you have forgotten to add `@return` Javadoc for this method. 
Was it something you accidentally neglected while refactoring for Level-10 JavaFX?

njxue replied:
Nice catch! No wonder you're an A+ student!

dexter-sim reviewed (commented):
Other than a minor Javadoc inconsistency, the code looks good.

njxue committed:
Add assert statements

njxue committed:
Convert task complettion status icons into constants

njxue committed:
Add new custom Duke exception called FileCorruptedException

There is no existing exception handling for cases where the storage file
containing the tasks are modified and cannnot be parsed properly.

Adding an exception that handles such a case allows the program to
continue running with an empty task list. Also, it allows for better
if-else code structure.

Let's create a new DukeException, FileCorruptedException, which extends
DukeException.

njxue committed:
Move index validity check into a new method

The code involved in checking where a task index is valid is duplicated
across multiple task instance methods.

Let's extract this code into another method to reduce code duplication.

njxue committed:
Extract splitting of full command into another method

njxue committed:
Reduce code duplication via Extract Method

njxue committed:
Move code block containing main logic to the top of the class

njxue committed:
Rename method to prefix with a verb

njxue committed:
Move checking of empty command to getCommandString method

njxue committed:
Move static fields to the top

njxue committed:
Update tests

njxue committed:
Merge pull request #1 from njxue/branch-A-CodeQuality

branch-A-CodeQuality

njxue committed:
Resolve merge conflicts

njxue committed:
Merge pull request #2 from njxue/branch-A-Assertions

branch-A-Assertions

kynapy reviewed (commented):
Code looks really good, really good OOP skills and JavaDocs clearly written. Impressive code, learnt a lot from this!

Charles1026 commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,45 @@
+     * @param tasks TaskList containing the list of tasks.
+     * @param ui Ui object which interacts with the user.
+     * @param storage Storage object which loads and saves tasks.
+     */
```
missing @return statement

MinHeinA reviewed (commented):
The packages are done neatly with the relevant classes placed in the same folder (package). Follows good OO practices, good naming of variables and well code is well documented.

Charles1026 commented on src/main/java/duke/gui/Ui.java:
```diff
@@ -0,0 +1,133 @@
+ * @version v0.1
+ */
+public class Ui {
+    private enum Divider {
```
maybe this could just be 2 class variables

Charles1026 commented on src/main/java/duke/gui/Ui.java:
```diff
@@ -0,0 +1,133 @@
+     * Returns the greeting message when the application is launched.
+     */
+    public String greetingMessage() {
+        String logo = "    __                    \n"
```
Nice logo 👍

Charles1026 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,121 @@
+     * @return Formatted task, which is to be written into the storage file.
+     */
+    public String toFileFormatString() {
+        String icon = isDone ? COMPLETED_ICON : INCOMPLETE_ICON;
```
can use getStatusIcon here

Charles1026 reviewed (commented):
👍

njxue committed:
Add sort by datetime functionality

njxue committed:
Move sorting of tasklist from TaskList to commands

njxue committed:
New feature: allow custom sorting of tasks in the TaskLists

TaskList is sorted according to deadline or event time and only in
ascending order. There is little flexibility in sorting.

Let's add more sorting capability to the TaskList.

Adding a Sort Command allows users to specify the order and metric to
sort the TaskList by.

njxue committed:
Fix style violations

njxue committed:
Move getCommandFormat method from Parse to Command class

njxue committed:
Merge pull request #3 from njxue/branch-C-Sort

branch-C-Sort

njxue committed:
Set up github workflows

njxue committed:
Refactor Parser: extract methods

The parse method, which houses the main logic of Parser, is too long and
verbose.

Extract the parsing procedure of each command into each method. This
improves code readability.

Let's
* Create a parse method to parse each Command
* Delete the getCommandFormat method from the Command class as Parser no
  longer requires it
* Use switch-case statements to improve readability

njxue committed:
Fix wrong error message

njxue committed:
Add guard clauses

njxue committed:
Fix bug and rename methods

njxue committed:
Rename sort order and metric

njxue committed:
Add padding and color around dialog boxes

njxue committed:
Add Reponse class

Success and error messages are non-differentiable. Thus, the GUI cannot
be set based on the nature of the response (success or error).

Let's:
* add a Response class to encapsulate a Reponse object that is
returned after a command executes
* add a Success response which extends the Response class
* add an Error response which extends the Response class

Using inheritance allows polymorphism and dynamic-binding of the isError
method.

njxue committed:
Merge branch 'master' into branch-A-BetterGui

njxue committed:
Improve Ui

njxue committed:
Improve Ui

njxue committed:
Change window dimensions and font size

njxue committed:
Fix incorrect command formats

njxue committed:
Add user guide and gui snippet

njxue committed:
Change font to make it compatible with Mac

njxue committed:
Add background image

njxue committed:
Update Ui screenshot

njxue committed:
Delete CLI tests

njxue committed:
Fix Ui issues

njxue committed:
Shorten divider length
